src/backend/admin_ops/users.py

```diff
@@ -0,0 +1,240 @@
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+from typing import Any, Dict, List, cast
+
+from pydantic import parse_obj_as
+
+from backend.admin_ops import do_request
+from backend.admin_ops.errors import MissingParameterError
+from backend.admin_ops.types import (
+    AuthUser,
+    UserInfo,
+    UserKeyOpParams,
+    UserKeys,
+    UserOpParams,
+    UserQuotaOpParams,
+    model_to_params,
+)
+
+
+async def get_user_info(
+    url: str,
+    access_key: str,
+    secret_key: str,
+    uid: str | None = None,
+    with_statistics: bool = False,
+) -> UserInfo:
+    """
+    Obtain informations about the user to whom the `access_key:secret_key` pair
+    belongs to.
+    """
+    params = {"access-key": access_key, "stats": with_statistics}
+    if uid is not None and len(uid) > 0:
+        params["uid"] = uid
+
+    res = await do_request(
+        url=url,
+        access_key=access_key,
+        secret_key=secret_key,
+        endpoint="/admin/user",
+        method="GET",
+        params=params,
+    )
+    return UserInfo.parse_obj(res.json())
+
+
+async def authenticate(url: str, access_key: str, secret_key: str) -> AuthUser:
+    """
+    Check if the given credentials are valid and whether the user is allowed
+    access to the admin ops API.
+    """
+    info = await get_user_info(url, access_key, secret_key)
+    return AuthUser(
+        user_id=info.user_id,
+        display_name=info.display_name,
+        is_admin=info.admin,
+    )
+
+
+async def list_uids(url: str, access_key: str, secret_key: str) -> List[str]:
+    """Obtain a list of all user ids."""
+
+    res = await do_request(
+        url=url,
+        access_key=access_key,
+        secret_key=secret_key,
+        method="GET",
+        endpoint="/admin/metadata/user",
+    )
+    lst: Any = res.json()
+    assert isinstance(lst, list)
+    return cast(List[str], lst)
+
+
+async def list_users(
+    url: str, access_key: str, secret_key: str, with_statistics: bool = False
+) -> List[UserInfo]:
+    """
+    Obtain a list of users, by first performing a `list_uids()` operation, and
+    then, for each `uid` returned, obtains said `uid` user info by running
+    `get_user_info()` in parallel. For a large enough number of `uids`, this
+    operation's complexity is non-negligible.
+    """
+
+    uids = await list_uids(url, access_key, secret_key)
+    assert len(uids) > 0
+
+    requests = [
+        get_user_info(
+            url,
+            access_key,
+            secret_key,
+            uid=uid,
+            with_statistics=with_statistics,
+        )
+        for uid in uids
```

Comment:
I admit my poor python knowledge ; what does this bit mean?
is it constructing an array using the get_user_info() as a sort-of lambda used in the subsequent for?

Reply:
It's building a list with the results of `get_info(url, access_key, secret_key, uid=uid, ...)`, for each `uid` in the list `uids`.

This is a fairly common python idiom to building lists, called "List Comprehension". More info can be found here: https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions